clients/src/test/java/org/apache/kafka/common/serialization/SerializationTest.java

```diff
@@ -106,6 +110,190 @@ public void stringSerdeShouldSupportDifferentEncodings() {
+        assertEquals(testData,
+            listSerde.deserializer().deserialize(topic, listSerde.serializer().serialize(topic, testData)),
+            "Should get the original collection of UUID after serialization and deserialization");
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void listSerdeSerializerShouldReturnByteArrayOfFixedSizeForUUIDInput() {
+        List<UUID> testData = Arrays.asList(UUID.randomUUID(), UUID.randomUUID(), UUID.randomUUID());
+        Serde<List<UUID>> listSerde = Serdes.ListSerde(ArrayList.class, Serdes.UUID());
+        assertEquals(117, listSerde.serializer().serialize(topic, testData).length,
+            "Should get length of 117 bytes after serialization");
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void listSerdeShouldRoundtripNonPrimitiveInput() {
+        List<String> testData = Arrays.asList("A", "B", "C");
+        Serde<List<String>> listSerde = Serdes.ListSerde(ArrayList.class, Serdes.String());
+        assertEquals(testData,
+            listSerde.deserializer().deserialize(topic, listSerde.serializer().serialize(topic, testData)),
+            "Should get the original collection of strings list after serialization and deserialization");
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void listSerdeShouldRoundtripPrimitiveInputWithNullEntries() {
+        List<Integer> testData = Arrays.asList(1, null, 3);
+        Serde<List<Integer>> listSerde = Serdes.ListSerde(ArrayList.class, Serdes.Integer());
+        assertEquals(testData,
+            listSerde.deserializer().deserialize(topic, listSerde.serializer().serialize(topic, testData)),
+            "Should get the original collection of integer primitives with null entries "
+                + "after serialization and deserialization");
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void listSerdeShouldRoundtripNonPrimitiveInputWithNullEntries() {
+        List<String> testData = Arrays.asList("A", null, "C");
+        Serde<List<String>> listSerde = Serdes.ListSerde(ArrayList.class, Serdes.String());
+        assertEquals(testData,
+            listSerde.deserializer().deserialize(topic, listSerde.serializer().serialize(topic, testData)),
+            "Should get the original collection of strings list with null entries "
+                + "after serialization and deserialization");
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void listSerdeShouldReturnLinkedList() {
+        List<Integer> testData = new LinkedList<>();
+        Serde<List<Integer>> listSerde = Serdes.ListSerde(LinkedList.class, Serdes.Integer());
+        assertTrue(listSerde.deserializer().deserialize(topic, listSerde.serializer().serialize(topic, testData))
+            instanceof LinkedList, "Should return List instance of type LinkedList");
+    }
+
+    @SuppressWarnings("unchecked")
```

Comment:
Thanks for the context, it is what it is (and I agree with your decision to prioritize ease of use). I was more wondering whether we might be missing a `SuppressWarnings("unchecked")` on one of the methods in the implementation, so that the user isn't forced to do the suppression themselves. But I can't quite tell where the warning is coming from, since it seems like we do already suppress unchecked warnings in `ListDeserializer#createListInstance`  where the casting occurs? Is it possible this was just left over from an earlier version, and we no longer need all the suppressions on these tests?

Reply:
No, unfortunately they are still needed.

AFAIK, suppression warnings cannot be propagated upwards.

In `ListDeserializer#createListInstance` we indeed using `SuppressWarnings("unchecked")` due to casting.
However, in tests it is used bc `Stack.class` (or `ArrayList.class`, `LinkedList.class`, etc) is a raw-type and does not guarantee the required inner type (`Integer` in this case). 

The only way to deal with these warnings is to create some wrapper classes as @mjsax suggested, like:
`public static class IntegerArrayList extends ArrayList<Integer> {}`
But I do not this it is a scalable and clean solution.

Reply:
Ah, I see, it's from the implicit casting of the parameters. That makes sense, I was just wondering since I didn't see any "obvious" casting in the test code itself. Thanks for the explanation